### PR TITLE
Changed polling mode to consume an event stream

### PR DESCRIPTION
**Pull request by mros123**
The not so ideal polling mode is changed to consume the event stream from OpenHAB.

### CODE REVIEW

File: src/openhab.cpp [hunk 1]
```diff
@@ -44,29 +44,128 @@ Integration* OpenHABPlugin::createIntegration(const QVariantMap& config, Entitie
...
+
+        // get item name from the topic string
+        // example: smarthome/items/EG_Esszimmer_Sonos_CurrentPlayingTime/state
+        QString name = map.value("topic").toString().split('/')[2];
+
+        EntityInterface* entity = m_entities->getEntityInterface(name);
+        if (entity != nullptr) {
+            qCDebug(m_logCategory) << "Debug: " + data;
+
+            QString value = ((QVariantMap*)(map.value("payload").data()))->value("value").toString();
```
Comment (martonborzak) [inline src/openhab.cpp]: Could you change the cast to?
`reinterpret_cast<QVariantMap*>(...)`

File: src/openhab.cpp [hunk 2]
```diff
@@ -44,29 +44,128 @@ Integration* OpenHABPlugin::createIntegration(const QVariantMap& config, Entitie
...
+            // because OpenHab doesn't send the item type in the status update, we have to extract it from our own
+            // entity library
+            if (entity->type() == "light") {
+                processLight(value, entity, false, false);
+            } else if (entity->type() == "blind") {
+                processBlind(value, entity);
+            }
+        } else if (_ohPlayerItems.contains(name)) {
+            QString value = ((QVariantMap*)(map.value("payload").data()))->value("value").toString();
```
Comment (martonborzak) [inline src/openhab.cpp]: Could you change the cast to?
`reinterpret_cast<QVariantMap*>(...)`

**Review by martonborzak — CHANGES_REQUESTED** [review]
Would be great to add the available entities from openhab to the available entities database with the `addAvailableEntity` method. So it would work with the web configurator, instead only with manual config. 

See more about the method here:
[https://github.com/YIO-Remote/integrations.library/blob/master/src/yio-plugin/integration.h#L65](https://github.com/YIO-Remote/integrations.library/blob/master/src/yio-plugin/integration.h#L65)

File: src/openhab.cpp (hunk 1 above)
Comment (mros123) [reply src/openhab.cpp]: It's changed.

File: src/openhab.cpp (hunk 2 above)
Comment (mros123) [reply src/openhab.cpp]: It's also changed.

File: src/openhab.cpp (hunk 1 above)
Comment (martonborzak) [reply src/openhab.cpp]: Thank you

File: src/openhab.cpp (hunk 2 above)
Comment (martonborzak) [reply src/openhab.cpp]: Thank you

**Comment by martonborzak** [comment]
cpplint still complains about this line:
`src/openhab.cpp:78:  Extra space before ( in function call  [whitespace/parens] [4]`
Could you check that one too? Thanks so much!

**Comment by mros123** [comment]
I will do this tomorrow and have a look to get cpplint to work locally. I will also integrate the available entity database.

File: src/openhab.cpp [hunk 3]
```diff
@@ -44,29 +44,130 @@ Integration* OpenHABPlugin::createIntegration(const QVariantMap& config, Entitie
...
+                processBlind(value, entity);
+            }
+        } else if (_ohPlayerItems.contains(name)) {
+            QString value =
+                (reinterpret_cast<const QVariantMap*>(map.value("payload").data()))->value("value").toString();
+            processPlayerItem(value, name);
+        }
+    }
+
+    if (_wasDisconnected == true) {
```
Comment (zehnm) [inline src/openhab.cpp]: `== true` part is redundant and should be omited

File: src/openhab.cpp [hunk 4]
```diff
@@ -83,18 +184,40 @@ void OpenHAB::connect() {
...
+    _sseReconnectTimer->stop();
+
     setState(DISCONNECTED);
 }
 
 void OpenHAB::enterStandby() {
-    _pollingTimer.setInterval(1000 * 60);  // 1 minute
+    _standby = true;
+
+    if (_sseReply->isRunning()) _sseReply->close();
```
Comment (zehnm) [inline src/openhab.cpp]: Please always use curly braces in conditions, even for single statements. It's better readable and prevents accidental errors.
I thought that check is included in cpplint :-)

File: src/openhab.cpp [hunk 5]
```diff
@@ -44,29 +44,130 @@ Integration* OpenHABPlugin::createIntegration(const QVariantMap& config, Entitie
...
     qCDebug(m_logCategory) << "setup";
 }
 
+void OpenHAB::streamReceived() {
+    QByteArray     rawData = _sseReply->readAll();
+    QByteArrayList splitted = rawData.split('\n');
+
+    foreach (QByteArray data, splitted) {
+        if (data.length() == 0) continue;
+        if (data.startsWith("event: message")) continue;
```
Comment (zehnm) [inline src/openhab.cpp]: Please combine both statements and use curly braces

File: src/openhab.cpp [hunk 6]
```diff
@@ -44,29 +44,130 @@ Integration* OpenHABPlugin::createIntegration(const QVariantMap& config, Entitie
...
     QObject::connect(&_nam, &QNetworkAccessManager::networkAccessibleChanged, this, &OpenHAB::onNetWorkAccessible);
 
     qCDebug(m_logCategory) << "setup";
 }
 
+void OpenHAB::streamReceived() {
+    QByteArray     rawData = _sseReply->readAll();
+    QByteArrayList splitted = rawData.split('\n');
+
+    foreach (QByteArray data, splitted) {
```
Comment (zehnm) [inline src/openhab.cpp]: `foreach` macro is discouraged since Qt 5.7 and will be removed in a future version:
https://doc.qt.io/qt-5/qtglobal.html#foreach

**Review by zehnm — CHANGES_REQUESTED** [review]
Thank you for your contribution!
Looks good, just a few cosmetic things.
Please change the deprecated `foreach` macro and the cpplint error, then it's ok to merge.
If you could add curly braces in all single condition statements, that would be great. I know this is a highly subjective topic and I thought cpplint checks for it...

**Review by zehnm — APPROVED** [review]

**Review by martonborzak — APPROVED** [review]

**Pull request merged**